# MetaMap NLP content view
This notebook intends to investigate the utility of MetaMap NLP content view (CVF=256) which claims to help NLP research. \
Here's a complete list of content views:
https://www.nlm.nih.gov/research/umls/knowledge_sources/metathesaurus/release/content_views.html

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
mrconso = pd.read_csv('/Users/qinyilong/Desktop/ScAi/MRCONSO.RRF', sep='|', header=None, dtype=object)

Meaning of columns: https://www.ncbi.nlm.nih.gov/books/NBK9685/table/ch03.T.concept_names_and_sources_file_mr/?report=objectonly

In [3]:
mrconso = mrconso.drop(18, axis=1) # last column is meaningless because the entry ends with '|'
mrconso.columns = ["CUI", "LAT", "TS", "LUI", "STT", "SUI", "ISPREF", "AUI", "SAUI", "SCUI", "SDUI", "SAB", "TTY", "CODE", "STR", "SRL", "SUPPRESS", "CVF"]

In [4]:
# Extract entries within MetaMap view
mrconso_metamap_on = mrconso[mrconso.CVF == "256"]

In [5]:
# Randomly sample 100 entries and display
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(mrconso_metamap_on.sample(100))

CUI  LAT TS        LUI STT        SUI ISPREF        AUI  \
7123265   C1189696  ENG  P   L2609809  PF   S3061078      N   A2736422   
12094084  C3494245  ENG  P  L10842059  VO  S13489904      Y  A20901751   
1974355   C0189573  ENG  P   L0228806  VO   S9046606      Y  A18652516   
7896375   C1412442  ENG  P   L5133263  PF   S5872488      Y   A6898457   
13450158  C4161913  ENG  P  L13192753  PF  S16174600      Y  A26725405   
4632034   C0524433  ENG  S   L0723345  PF   S0945647      N   A1001963   
10581504  C2641837  ENG  P   L8630787  PF  S10735414      Y  A16587703   
12876809  C3876513  ENG  P  L11803691  PF  S14674327      Y  A24240783   
8073304   C1445505  ENG  P  L13103163  PF  S16056918      Y  A26574894   
780821    C0026850  ENG  S   L1412288  VO  S11916484      Y  A18648618   
14040530  C4507950  ENG  P  L14167661  PF  S17236230      Y  A28393524   
12200393  C3537590  ENG  P  L11041051  PF  S13746742      Y  A21258029   
13041587  C3949829  ENG  P  L11907864  PF  S14913659      Y  A24410496   
12578980  C3728497  ENG  S  L11539995  PF  S14340853      Y  A23675240   
10938784  C2802216  ENG  P   L9249596  PF  S11459861      Y  A17495508   
1115129   C0038683  ENG  S  L13415237  PF  S16395261      Y  A27411113   
1798543   C0159877  ENG  P   L0186772  VO  S14003057      Y  A22790895   
131382    C0005048  ENG  S   L0005582  PF   S0019298      Y  A26614118   
13936854  C4444457  ENG  S  L14045674  PF  S16940788      Y  A28166595   
2259474   C0224483  ENG  S   L0287793  PF   S0373058      N  A27034272   
8560849   C1621993  ENG  S  L10779123  PF  S13400976      Y  A20634127   
2476919   C0241405  ENG  P   L0276866  PF   S0408267      Y   A0464319   
8240897   C1514789  ENG  P   L5354311  PF   S6140062      Y   A7660730   
1883251   C0175649  ENG  P   L0033584  PF   S0077298      N   A0106217   
690264    C0023690  ENG  S   L1058029  PF   S1273345      N   A4867143   
6786054   C1070642  ENG  S   L2023125  PF   S2325170      Y   A2174846   
11941113  C3383943  ENG  P  L10601548  PF  S13067790      Y  A20421009   
1312518   C0056889  ENG  P   L0056889  PF   S0479611      N   A0546410   
12694432  C3811964  ENG  P  L11614143  PF  S14406913      N  A23816521   
11251859  C2931317  ENG  P   L0159828  VO   S0506527      Y  A27369400   
969801    C0033779  ENG  P   L0033779  VO   S0080134      Y   A0109635   
3211081   C0336188  ENG  P   L0568598  PF   S0652725      N   A0708537   
6571423   C0981136  ENG  S   L1903653  PF   S2248756      N  A17939074   
13374943  C4101098  ENG  P  L13168522  PF  S16168675      Y  A26721611   
3490929   C0366250  ENG  S   L2426267  PF   S2859489      Y  A18307851   
10222804  C2385001  ENG  P   L8268832  PF  S10487548      N  A19803337   
13733428  C4309592  ENG  P  L13788713  PF  S16812832      Y  A27811462   
12495820  C3686708  ENG  S  L11342756  VO  S14083672      Y  A23163768   
13729305  C4307371  ENG  S  L13786666  PF  S16815019      Y  A27811974   
8134972   C1467326  ENG  P   L6599173  PF   S7712260      Y  A12887460   
12733091  C3824332  ENG  S  L11641154  PF  S14428502      Y  A23853856   
5838967   C0833415  ENG  P   L1663238  PF   S1874011      Y   A1819814   
13353189  C4091246  ENG  P  L13174004  PF  S16211817      Y  A26905250   
4808266   C0559408  ENG  P   L1352354  PF   S1601181      Y   A3314649   
13454604  C4166267  ENG  P  L13304637  PF  S16139126      Y  A26867204   
1449769   C0079931  ENG  S   L0079979  VO   S0070304      N  A17978273   
2846784   C0282537  ENG  S   L0404688  VO  S11948165      Y  A18692356   
9118035   C1926389  ENG  P   L6658724  PF   S7691219      Y  A12918047   
1988223   C0191017  ENG  P   L0230751  VO   S0305921      Y   A0336687   
322470    C0011816  ENG  S  L13415325  PF  S16396183      Y  A27057356   
2145500   C0205491  ENG  P   L0249147  VO  S11922121      Y  A18652887   
6922193   C1136058  ENG  S   L1224162  VO  S14443630      Y  A23872390   
12549778  C3713420  ENG  P  L11461818  PF  S14249249      N  A27940331   
11907452  C3353621 

In [6]:
# Frequencies of entries in different languages
mrconso_metamap_on["LAT"].value_counts()

ENG    4802688
Name: LAT, dtype: int64

In [7]:
sources_metamap_on = mrconso_metamap_on["SAB"].value_counts()

In [8]:
# Extract entries without MetaMap view
mrconso_metamap_off = mrconso[mrconso.CVF != "256"]

In [9]:
# Randomly sample 100 entries and display
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(mrconso_metamap_off.sample(100))

CUI  LAT TS        LUI STT        SUI ISPREF        AUI  \
9481806   C2065223  ENG  S   L7273125  PF   S8730711      Y  A13892990   
12431412  C3662551  ENG  P  L14906876  VO  S18100673      Y  A29536913   
8945733   C1831490  EST  P  L12669192  PF  S15126651      Y  A25178532   
3574148   C0376633  DUT  P   L2051275  PF   S2389497      Y   A8182762   
7207959   C1245153  ENG  P   L2656218  PF   S3123690      Y  A10492711   
10144214  C2350514  RUS  S   L8589020  PF  S10683676      Y  A16462733   
9361605   C2022698  ENG  P   L9004393  PF  S11200520      Y  A17264186   
6322469   C0941961  KOR  P  L12469211  PF  S15869447      Y  A25420544   
2044581   C0195913  ENG  S   L2884274  PF   S3304076      Y   A3442836   
5306986   C0693979  SPA  S  L14344087  PF  S17437578      Y  A28661409   
1814757   C0160746  ENG  P   L0187641  VO   S4093403      Y   A4426025   
9700802   C2145939  ENG  S   L7867143  PF   S9200870      Y  A15285634   
10847563  C2747821  JPN  P  L11209171  PF  S13929997      N  A26438108   
2651283   C0267953  GER  P   L5039385  PF   S5726890      N  A25866838   
10813360  C2735466  DUT  P  L14492177  PF  S17584815      Y  A28835263   
7827190   C1393553  ENG  P   L3528556  VO   S4066505      Y   A4399127   
13352247  C4090823  ENG  S  L13214957  PF  S16166771      Y  A26855536   
9283446   C1983370  ENG  P  L11246331  PF  S13967559      Y  A22776800   
9785641   C2177637  ENG  S   L8803756  PF  S11007125      Y  A16879317   
1238312   C0043043  POL  P   L9678275  PF  S12152417      Y  A18998860   
3857666   C0419627  SPA  S   L4202578  PF   S4886047      Y  A29311472   
3717461   C0404572  SPA  S   L4327313  PF   S5010781      N  A23188294   
9036440   C1875852  ENG  P   L6532397  PF   S7615433      N  A18033594   
4331296   C0477355  SPA  S   L4927948  PF   S5615453      Y  A11333026   
6085403   C0872271  SWE  P   L3430736  PF   S3958462      Y  A11750670   
8654496   C1660793  ENG  P   L5817331  PF   S6651912      Y   A9336927   
8802333   C1717187  POR  P  L12287722  PF  S15195426      Y  A25233856   
8056691   C1442861  ENG  P   L0005676  PF   S0002989      Y   A2875005   
1304041   C0055530  ITA  S   L8136015  PF  S10184351      Y  A27518940   
12593581  C3740207  ENG  P  L11475040  PF  S14318104      Y  A23677127   
7082419   C1175140  ENG  S   L2504844  PF   S2969051      Y   A2627722   
9774983   C2173501  ENG  S   L8813224  PF  S11011975      Y  A16823385   
8464629   C1564995  ENG  P   L5556232  VO   S6354246      Y   A8407900   
5510759   C0751713  GER  P   L3285272  PF   S3812908      Y  A13466804   
11395356  C2987188  ENG  P   L9716774  VO  S18095894      Y  A29480081   
2361937   C0231397  SPA  S   L4919414  PF   S5606919      Y   A6517890   
13250437  C4056932  ENG  S  L13002982  PF  S15945255      Y  A26447931   
13371483  C4099537  ENG  S  L13320369  PF  S16265860      Y  A26795777   
9223453   C1963843  CZE  P  L14773205  PF  S17976539      Y  A29328757   
3924962   C0426265  POR  S   L4912099  PF   S5599605      N  A26383702   
10819405  C2736051  ENG  S  L10953634  PF  S13645390      Y  A21121766   
9658658   C2129375  ENG  P   L7346492  PF   S8708728      Y  A14078049   
1555024   C0123538  FRE  S   L3270911  PF   S3800258      Y  A27471616   
14145867  C4537991  ENG  P  L14403876  PF  S17505363      Y  A28755515   
2959525   C0312684  SPA  S   L3891130  PF   S4574594      Y  A12529807   
10865235  C2753387  ENG  S  L10434647  PF  S13004353      Y  A20227437   
4999204   C0583680  SPA  S   L4837694  PF   S5521163      Y   A6305066   
9322747   C2008146  ENG  P   L7053755  PF   S8139884      Y  A14070978   
1051255   C0036508  ENG  P   L0011612  VO   S1634898      N   A4375993   
6026258   C0860225  CZE  P   L8024302  PF  S10096309      Y  A26330129   
14234232  C4575908  ENG  P  L14708050  PF  S17700864      Y  A29107731   
6492322   C0975652  ENG  S   L5568026  PF   S6386285      N  A18008582   
11845411  C3274282  ENG  S  L10438415  PF  S13026244      Y  A20240662   
8885403   C1821942 

In [10]:
# Frequencies of entries in different languages
mrconso_metamap_off["LAT"].value_counts()

ENG    5508879
SPA    1470221
FRE     424471
JPN     322983
POR     300417
DUT     286526
ITA     236976
GER     232823
RUS     180945
CZE     180525
HUN     104811
CHI      77622
NOR      61724
TUR      51114
POL      50108
KOR      38660
EST      31204
SWE      29737
FIN      25486
SCR       9981
GRE       2241
LAV       1405
DAN        723
BAQ        695
HEB        485
Name: LAT, dtype: int64

In [11]:
sources_metamap_off = mrconso_metamap_off["SAB"].value_counts()

In [12]:
# Let's what english words are excluded from MetaMap view
mrconso_metamap_off_eng = mrconso_metamap_off[mrconso_metamap_off.LAT == "ENG"]

In [13]:
# Randomly sample 100 entries and display
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(mrconso_metamap_off_eng.sample(100))

CUI  LAT TS        LUI STT        SUI ISPREF        AUI  \
8529859   C1609327  ENG  P   L5642663  PF   S6470341      Y   A8600695   
2054454   C0196965  ENG  S   L2874560  PF   S3263890      Y   A3401147   
39827     C0001911  ENG  S   L2811869  PF  S18100794      Y  A29558076   
11756528  C3246195  ENG  S  L10251028  PF  S12791751      Y  A19864400   
702025    C0024032  ENG  S   L6193676  VO   S0814224      Y   A0873907   
2612794   C0265840  ENG  S   L0550656  PF   S0743230      Y   A0804119   
13983010  C4480320  ENG  P  L14090591  VO  S17139225      Y  A28278445   
13264674  C4065413  ENG  S  L13027668  PF  S15964929      Y  A26474389   
11370215  C2977641  ENG  P  L10347029  PF  S12929449      Y  A20142003   
5419800   C0726736  ENG  P   L1333837  PF   S1577502      Y   A9431202   
13646202  C4281884  ENG  S  L13647774  PF  S17929736      Y  A29227651   
14338368  C4662490  ENG  P  L14641730  PF  S17691243      Y  A29025953   
14303633  C4630520  ENG  P  L14602004  PF  S17719994      Y  A29104307   
5106929   C0606599  ENG  P   L1141943  PF   S1369869      Y   A1330190   
4878508   C0568168  ENG  S   L7896803  PF   S9198029      Y  A15235749   
12532313  C3702513  ENG  P  L11448321  PF  S15894903      Y  A25757053   
9709168   C2149060  ENG  P   L7233705  PF   S8387690      Y  A14163299   
5929994   C0854313  ENG  S   L7872690  PF   S9086042      Y  A15222986   
12761180  C3835384  ENG  S  L11697779  PF  S14534399      Y  A24066230   
14001446  C4488732  ENG  P  L14096270  PF  S17168803      Y  A28286976   
4885988   C0569055  ENG  S   L7077183  PF   S8337044      Y  A13873637   
7570487   C1306551  ENG  P   L3137448  PF   S3660354      Y  A10503657   
1963905   C0188605  ENG  P   L2822013  VO   S3433132      Y  A22916497   
1435499   C0079152  ENG  P   L0079152  VO  S11870769      Y  A18669564   
4356280   C0478863  ENG  S   L0858078  PF   S0913236      Y   A0969863   
11347626  C2969934  ENG  S   L6428013  PF  S12603867      Y  A19508039   
2957559   C0312389  ENG  P   L0498347  PF   S0626790      N   A0682823   
5418251   C0725321  ENG  S   L7654020  PF   S8866461      Y  A17981030   
10919225  C2787998  ENG  S   L9206665  PF  S11439955      Y  A17578408   
10400871  C2498634  ENG  P   L8373742  PF  S10428928      Y  A19772688   
4789805   C0557363  ENG  S   L3046825  PF   S3508440      Y   A3663468   
2076992   C0199166  ENG  S  L14216661  PF  S17280321      Y  A28402477   
12812380  C3854167  ENG  S  L11779822  PF  S14639232      Y  A24212082   
13963633  C4465197  ENG  S  L13875055  PF  S16925979      Y  A28152587   
7633050   C1318017  ENG  S   L2917929  PF   S3196046      Y   A3326467   
11518935  C3091883  ENG  P   L9768523  PF  S12293629      Y  A19235054   
13391508  C4108448  ENG  S  L13169864  PF  S16179171      Y  A26733296   
5852697   C0839470  ENG  P   L1690258  VO  S17143737      Y  A28272396   
14340288  C4664410  ENG  P  L14707288  PF  S17694999      Y  A29065752   
12153182  C3515639  ENG  P  L10912928  PF  S13583335      N  A23607321   
11193632  C2908972  ENG  S  L14941934  PF  S18188643      Y  A29593019   
11611475  C3161492  ENG  P   L9986970  PF  S12495942      Y  A19387199   
11064475  C2858248  ENG  P   L9351203  PF  S11660036      N  A17793996   
5260155   C0688786  ENG  P   L1316909  PF   S2679009      Y  A10420552   
9069512   C1891948  ENG  S   L6657366  PF   S7680359      Y  A28091735   
10194670  C2365903  ENG  S   L9421927  PF  S11718784      Y  A17914660   
5305713   C0693823  ENG  S   L7667062  PF   S8868192      Y  A17956348   
3669473   C0399220  ENG  P   L0877601  PF   S0976125      Y   A1034486   
10231032  C2391361  ENG  S   L8221224  PF  S10406164      Y  A16315527   
10665548  C2693964  ENG  S  L10253944  PF  S12816223      Y  A19786352   
11742801  C3232839  ENG  P  L10191525  PF  S13399245      Y  A20625426   
11377969  C2981013  ENG  S  L10934399  PF  S13618504      Y  A28214718   
7875425   C1406752  ENG  P   L3546867  PF   S4127299      Y   A4459921   
6809897   C1083222 

In [14]:
sources = pd.DataFrame(dict(included = sources_metamap_on, excluded = sources_metamap_off))

In [15]:
# Frequencies of entries from different source vocabularies
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(sources)

included  excluded
AIR                  337.0       347
ALT                  200.0      8865
AOD                17933.0      2746
AOT                  444.0        27
ATC                 5506.0      1310
BI                   951.0       300
CCC                   62.0       348
CCPSS               1764.0     14070
CCS                  936.0       681
CCS_10               444.0       436
CDT                   13.0       785
CHV               125619.0     20533
COSTAR              3339.0       119
CPM                  446.0      2649
CPT                 2557.0     68333
CPTSP                  NaN      2684
CSP                21674.0      1093
CST                 4263.0      2179
CVX                  234.0       389
DDB                   29.0       227
DMDICD10               NaN     11999
DMDUMD                 NaN      3339
DRUGBANK           15726.0      7454
DSM-5                236.0      1101
DXP                 9109.0      1003
FMA               156994.0     14003
GO                118786.0     60134
GS                   570.0     61066
HCDT                  32.0      1378
HCPCS                878.0     12300
HCPT                 405.0     20742
HGNC              154424.0     19307
HL7V2.5             3114.0      1905
HL7V3.0             9560.0      2442
HPO                27221.0      3552
ICD10               5571.0      7929
ICD10AE              459.0       648
ICD10AM             7517.0     18364
ICD10AMAE            749.0      1656
ICD10CM            38516.0    140534
ICD10DUT               NaN     11359
ICD10PCS           12992.0    336914
ICD9CM             14635.0     26178
ICF                  730.0       791
ICF-CY               825.0       946
ICNP                 178.0      1863
ICPC                 640.0       413
ICPC2EDUT              NaN       685
ICPC2EENG            542.0       837
ICPC2ICD10DUT          NaN     76614
ICPC2ICD10ENG       2323.0     79451
ICPC2P              4765.0     12124
ICPCBAQ                NaN       695
ICPCDAN                NaN       723
ICPCDUT                NaN       723
ICPCFIN                NaN       722
ICPCFRE                NaN       723
ICPCGER                NaN       723
ICPCHEB                NaN       485
ICPCHUN                NaN       718
ICPCITA                NaN       723
ICPCNOR                NaN       722
ICPCPOR                NaN       723
ICPCSPA                NaN       723
ICPCSWE                NaN       723
JABL                 530.0      2730
KCD5                   NaN     10997
LCH                 6407.0       244
LCH_NW             12533.0       795
LNC               203805.0    275488
LNC-DE-AT              NaN      5814
LNC-DE-CH              NaN      4941
LNC-DE-DE              NaN     11059
LNC-EL-GR              NaN      2241
LNC-ES-AR              NaN     76586
LNC-ES-CH              NaN      4940
LNC-ES-ES              NaN     54206
LNC-ET-EE              NaN     31204
LNC-FR-BE              NaN     45905
LNC-FR-CA              NaN     45349
LNC-FR-CH              NaN      4940
LNC-FR-FR              NaN     96894
LNC-IT-CH              NaN      4939
LNC-IT-IT              NaN     66396
LNC-KO-KR              NaN     26893
LNC-NL-NL              NaN     52516
LNC-PT-BR              NaN    110887
LNC-RU-RU              NaN     57499
LNC-TR-TR              NaN     51114
LNC-ZH-CN              NaN     77622
MCM                   39.0         4
MDR                29717.0     75619
MDRCZE                 NaN    104092
MDRDUT                 NaN    104092
MDRFRE                 NaN    104092
MDRGER                 NaN    104092
MDRHUN                 NaN    104093
MDRITA                 NaN    104094
MDRJPN                 NaN    198936
MDRPOR                 NaN    104094
MDRSPA                 NaN    104094
MED-RT              6630.0      1213
MEDCIN             21483.0    879813
MEDLINEPLUS         2986.0       393
MMSL               10877.0     71266
MMX                 1513.0     71119
MSH               609456.0    262484
MSHCZE                

In [16]:
# Some shapes
print("Entire MRCONSO: {}".format(mrconso.shape))
print("MetaMap view: {}".format(mrconso_metamap_on.shape))
print("MetaMap view excluded: {}".format(mrconso_metamap_off.shape))
print("English entries excluded by MetaMap view: {}".format(mrconso_metamap_off_eng.shape))

Entire MRCONSO: (14433450, 18)
MetaMap view: (4802688, 18)
MetaMap view excluded: (9630762, 18)
English entries excluded by MetaMap view: (5508879, 18)


### Observations:
Metamap NLP content view extracts 4802688 english concepts and discards 5508879 english concepts and all the non-english concepts. \
Source Vocabularies like LONIC (LNC-), which are suggested by reference manual to be detrimental to NLP, are completely excluded. \

### TODO:
What kinds of English concepts were excluded?

#### Check if MetaMap helps to reduce mappings from one CUI to multiple STYs 

In [23]:
import pickle
with open('pickles/STY_to_CUIs.pickle', 'rb') as handle:
    sty_to_cuis = pickle.load(handle)

In [27]:
cui_to_stys_metamap = dict((el, [])for el in mrconso_metamap_on.CUI.drop_duplicates())
for sty in sty_to_cuis:
    for cui in sty_to_cuis[sty]:
        if cui in cui_to_stys_metamap:
            cui_to_stys_metamap[cui] += [sty]

In [29]:
from collections import Counter
ctr = Counter([len(x) for x in cui_to_stys_metamap.values()])
ctr.most_common(10)

[(1, 1975275), (2, 210001), (3, 8652), (4, 45)]

In [30]:
with open('pickles/CUI_to_STYs.pickle', 'rb') as handle:
    cui_to_stys = pickle.load(handle)

In [31]:
ctr1 = Counter([len(x) for x in cui_to_stys.values()])
ctr1.most_common(10)

[(1, 3538314), (2, 271362), (3, 13088), (4, 68)]

#### Check if Metamap affects the integrity of tree structures in MRHIER

In [32]:
mrhier = pd.read_csv('/Users/qinyilong/Desktop/ScAi/MRHIER.RRF', sep='|', header=None, dtype=object)
mrhier = mrhier.drop(9, axis=1)
mrhier.columns = ['CUI', 'AUI', 'CXN', 'PAUI', 'SAB', 'RELA', 'PTR', 'HCD', 'CVF']

In [33]:
mrhier_ncbi = mrhier[mrhier['SAB'] == 'NCBI']
mrhier_ncbi = mrhier_ncbi[mrhier_ncbi['PTR'].notna()]

In [44]:
import networkx as nx
G = nx.DiGraph(name='MRHIER NCBI hierarchy graph')
for index, row in mrhier_ncbi.iterrows():
    parents = row['PTR'].split('.')
    for index, aui in enumerate(parents):
        if aui in mrconso_metamap_on.AUI and parents[index+1] in mrconso_metamap_on.AUI:
            G.add_edge(aui, parents[index+1])
    if parents[-1] in mrconso_metamap_on.AUI and row['AUI'] in mrconso_metamap_on.AUI:
        G.add_edge(parents[-1], row['AUI'])

In [45]:
nx.is_tree(G)

NetworkXPointlessConcept: G has no nodes.

In [46]:
nx.info(G)

'Name: MRHIER NCBI hierarchy graph\nType: DiGraph\nNumber of nodes: 0\nNumber of edges: 0\n'